File: stripe/src/main/java/com/stripe/android/PaymentConfiguration.java
package com.stripe.android;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.stripe.android.utils.ObjectUtils;

import java.util.Objects;

public final class PaymentConfiguration implements Parcelable {

    @Nullable private static PaymentConfiguration mInstance;
    @NonNull private final String mPublishableKey;

    private PaymentConfiguration(@NonNull String publishableKey) {
        mPublishableKey = ApiKeyValidator.get().requireValid(publishableKey);
    }

    private PaymentConfiguration(@NonNull Parcel in) {
        mPublishableKey = Objects.requireNonNull(in.readString());
    }

    public static final Creator<PaymentConfiguration> CREATOR =
            new Creator<PaymentConfiguration>() {
                @Override
                public PaymentConfiguration createFromParcel(@NonNull Parcel in) {
                    return new PaymentConfiguration(in);
                }

                @Override
                public PaymentConfiguration[] newArray(int size) {
                    return new PaymentConfiguration[size];
                }
            };

    /**
     * Attempts to load a {@link PaymentConfiguration} instance. First attempt to use the class's
     * singleton instance. If unavailable, attempt to load from {@link Store}.
     *
     * @param context application context
     * @return a {@link PaymentConfiguration} instance, or throw an exception
     */
    @NonNull
    public static PaymentConfiguration getInstance(@NonNull Context context) {
        if (mInstance == null) {
            final PaymentConfiguration loadedInstance = new Store(context).load();
            if (loadedInstance != null) {
                mInstance = loadedInstance;
            } else {
                throw new IllegalStateException("PaymentConfiguration was not initialized");
            }
        }
        return mInstance;
    }

    /**
     * A publishable key from the Dashboard's
     * <a href="https://dashboard.stripe.com/apikeys">API keys</a> page.
     */
    public static void init(@NonNull Context context, @NonNull String publishableKey) {
        mInstance = new PaymentConfiguration(publishableKey);
        new Store(context).save(publishableKey);
    }

    @NonNull
    public String getPublishableKey() {
        return mPublishableKey;
    }

    @VisibleForTesting
    static void clearInstance() {
        mInstance = null;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(mPublishableKey);
    }

    @Override
    public int hashCode() {
        return Objects.hash(mPublishableKey);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        return super.equals(obj) ||
                (obj instanceof PaymentConfiguration && typedEquals((PaymentConfiguration) obj));
    }

    private boolean typedEquals(@NonNull PaymentConfiguration obj) {
        return Objects.equals(mPublishableKey, obj.mPublishableKey);
    }

    /**
     * Manages saving and loading {@link PaymentConfiguration} data to SharedPreferences.
     */
    private static final class Store {
        @NonNull private final SharedPreferences mPrefs;
        private static final String NAME = PaymentConfiguration.class.getCanonicalName();

        private static final String KEY_PUBLISHABLE_KEY = "key_publishable_key";

        private Store(@NonNull Context context) {
            mPrefs = context.getApplicationContext().getSharedPreferences(NAME, 0);
        }

        private void save(@NonNull String publishableKey) {
            mPrefs.edit()
                    .putString(KEY_PUBLISHABLE_KEY, publishableKey)
                    .apply();
        }

        @Nullable
        private PaymentConfiguration load() {
            final String publishableKey = mPrefs.getString(KEY_PUBLISHABLE_KEY, null);
            if (publishableKey == null) {
                return null;
            }
            return new PaymentConfiguration(publishableKey);
        }
    }
}

File: stripe/src/main/java/com/stripe/android/PaymentConfiguration.kt
package com.stripe.android

import android.content.Context
import android.content.SharedPreferences
import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

@Parcelize
data class PaymentConfiguration internal constructor(
    val publishableKey: String,
    val stripeAccountId: String? = null,
    val betas: Set<StripeApiBeta> = emptySet()
) : Parcelable {

    init {
        ApiKeyValidator.get().requireValid(publishableKey)
    }

    /**
     * Manages saving and loading [PaymentConfiguration] data to SharedPreferences.
     */
    private class Store internal constructor(context: Context) {
        private val prefs: SharedPreferences =
            context.applicationContext.getSharedPreferences(NAME, 0)

        @JvmSynthetic
        internal fun save(
            publishableKey: String,
            stripeAccountId: String?,
            betas: Set<StripeApiBeta>
        ) {
            prefs.edit()
                .putString(KEY_PUBLISHABLE_KEY, publishableKey)
                .putString(KEY_ACCOUNT_ID, stripeAccountId)
                .putStringSet(KEY_BETAS, betas.mapTo(mutableSetOf(), { it.code }))
                .apply()
        }

        @JvmSynthetic
        internal fun load(): PaymentConfiguration? {
            return prefs.getString(KEY_PUBLISHABLE_KEY, null)?.let { publishableKey ->
                PaymentConfiguration(
                    publishableKey = publishableKey,
                    stripeAccountId = prefs.getString(KEY_ACCOUNT_ID, null),
                    betas = prefs.getStringSet(KEY_BETAS, emptySet())
                        .orEmpty()
                        .mapNotNull { StripeApiBeta.fromCode(it) }
                        .toSet()
                )
            }
        }

        private companion object {
            private val NAME = PaymentConfiguration::class.java.canonicalName

            private const val KEY_PUBLISHABLE_KEY = "key_publishable_key"
            private const val KEY_ACCOUNT_ID = "key_account_id"
            private const val KEY_BETAS = "key_betas"
        }
    }

    companion object {
        private var instance: PaymentConfiguration? = null

        /**
         * Attempts to load a [PaymentConfiguration] instance. First attempt to use the class's
         * singleton instance. If unavailable, attempt to load from [Store].
         *
         * @param context application context
         * @return a [PaymentConfiguration] instance, or throw an exception
         */
        @JvmStatic
        fun getInstance(context: Context): PaymentConfiguration {
            return instance ?: loadInstance(context)
        }

        private fun loadInstance(context: Context): PaymentConfiguration {
            return Store(context).load()?.let {
                instance = it
                it
            }
                ?: throw IllegalStateException(
                    "PaymentConfiguration was not initialized. Call PaymentConfiguration.init()."
                )
        }

        /**
         * A publishable key from the Dashboard's [API keys](https://dashboard.stripe.com/apikeys) page.
         */
        @JvmStatic
        @JvmOverloads
        fun init(
            context: Context,
            publishableKey: String,
            stripeAccountId: String? = null,
            betas: Set<StripeApiBeta> = emptySet()
        ) {
            instance = PaymentConfiguration(
                publishableKey = publishableKey,
                stripeAccountId = stripeAccountId,
                betas = betas
            )
            Store(context)
                .save(
                    publishableKey = publishableKey,
                    stripeAccountId = stripeAccountId,
                    betas = betas
                )

            FingerprintDataRepository.Default(context).refresh()
        }

        @JvmSynthetic
        internal fun clearInstance() {
            instance = null
        }
    }
}
